### PR TITLE
fix: recover scoped vector matches beyond initial KNN window

### LINKED ISSUE #100: Chunk-KNN date/scope filters applied after top_k limiting, can miss in-range matches
## Problem

`execute_chunk_knn` in `src/ccrecall/search_vector.py` runs the vec MATCH query for the globally-nearest `top_k` chunks first, then applies `projects`/`session_id`/`path`/`before`/`after` filters afterward on that already-limited candidate set:

```python
knn_rows = cursor.execute(
    "SELECT chunk_id, distance FROM chunk_vec WHERE embedding MATCH ? AND k = ? ORDER BY distance",
    (serialized, top_k),
).fetchall()
# ... scope filters (project/session/path/date) applied only to these top_k rows
```

When a scope filter is selective (e.g. a narrow `--before`/`--after` range, or a project with few matching chunks), the truly relevant in-range chunks can rank below the global `top_k` cutoff and never reach the filter step — `search-messages` can return an empty or underfilled result set despite matching chunks existing, and `search`'s semantic rung loses the same candidates.

## Origin

Flagged by automated review (`chatgpt-codex-connector`) on PR #99 (`search_vector.py:60`). Confirmed **pre-existing**, not introduced by that PR — the same unconstrained-KNN-then-filter pattern already existed for `projects`/`session_id`/`path` before `--before`/`--after` was added; PR #99 just extended the existing pattern to dates.

## Proposed fix (needs design decision, not a one-liner)

Options to evaluate:
- Adaptive overfetch: widen `top_k` (or loop, retrying with a larger k) when any scope filter is active, scaled to filter selectivity
- sqlite-vec partition/metadata columns on `chunk_vec` to push scope predicates into the KNN query itself, avoiding post-filtering entirely

Whatever approach is chosen should fix this for all four scope filters together (project/session/path/date), not just dates, since they share the same root cause.

## Current overfetch constants (for context)

`search_conversations.py`: `OVERFETCH_MULTIPLIER = 4`, `OVERFETCH_FLOOR = 20`, `CHUNK_COLLAPSE_FACTOR = 8` — fixed heuristics, not filter-selectivity-aware.

### PR DESCRIPTION
**Pull request by NodeJSmith**
### Search Recall
- Retry chunk-vector KNN with larger windows when project, session, path, date, or currentness filters underfill the first pass so scoped searches recover valid farther matches instead of returning false empty results.
- Keep both search entrypoints on the same vector core while letting message search ask for enough filtered snippet hits, preserving distance ordering and the existing narrow sqlite degradation behavior.

### Test Coverage
- Add adversarial vector-search regressions for scope filters, stale or inactive chunks, Track A fusion, Track B snippet ordering, lazy count behavior, and the new sqlite error boundary so the retry path is pinned from multiple angles.

### Housekeeping
- Capture the investigation and design notes for this work in `design/research/2026-08-03-chunk-knn-scope-filtering/research.md` and `design/specs/006-chunk-knn-scope-filtering/design.md`.

Closes #100

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved filtered vector search so valid matches beyond the initial result window are now discovered.
  * Preserved result ordering while excluding stale, invalid, or out-of-scope matches.
  * Improved handling of date, project, session, and path filters.
  * Maintained graceful behavior for database errors and invalid search requests.

* **Documentation**
  * Added research and design documentation describing the search limitation, solution, acceptance criteria, and edge cases.
  * Added a changelog entry for the improved scoped vector retrieval.

* **Tests**
  * Added comprehensive coverage for adaptive retrieval, filtering, fallback behavior, and error handling.

### CODE REVIEW

File: src/ccrecall/search_vector.py [hunk 1]
```diff
@@ -7,6 +7,94 @@
...
+    path: str | None,
+    before: str | None,
+    after: str | None,
+) -> list[tuple[int, int, float]]:
+    if not knn_rows:
+        return []
+
+    chunk_ids = [row[0] for row in knn_rows]
+    chunk_to_dist: dict[int, float] = {row[0]: row[1] for row in knn_rows}
+    placeholders = ",".join("?" * len(chunk_ids))
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/ccrecall/search_vector.py]: Fixed — chunk IDs are now filtered in bounded batches, preserving KNN order while accounting for scope/currentness parameters in the SQL bind budget.

File: src/ccrecall/search_vector.py [hunk 2]
```diff
@@ -7,6 +7,94 @@
...
+    )
+    filter_sql = (
+        f"SELECT ch.id as chunk_id, b.id as branch_id FROM chunks ch {joins_sql}"
+        f"WHERE ch.id IN ({placeholders}) {where_sql}"
+    )
+    filter_params: list = [*chunk_ids, *filter_tail_params]
+
+    valid_rows = cursor.execute(filter_sql, filter_params).fetchall()
+    chunk_to_branch: dict[int, int] = {row[0]: row[1] for row in valid_rows}
+    return [(cid, chunk_to_branch[cid], chunk_to_dist[cid]) for cid in chunk_ids if cid in chunk_to_branch]
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/ccrecall/search_vector.py:85]: Fixed — chunk IDs are now batched before relational filtering, with regression coverage for batched filtering and exhausted parameter-budget handling.

File: design/research/2026-08-03-chunk-knn-scope-filtering/research.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+---
+proposal: "Fix ccrecall vector search so project/session/path/date filters cannot exclude valid in-scope chunks before they are considered."
+date: 2026-08-03
+status: Draft
+flexibility: Exploring
+motivation: "Best recall. Prioritize search correctness even if the implementation is more involved."
+constraints: "Scope is the search subsystem; inspect related search contracts and CLI behavior, not a full rewrite. No hard constraints."
+non-goals: "No full search rewrite; no application-code changes in this research pass."
+depth: normal
+---
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply design/research/2026-08-03-chunk-knn-scope-filtering/research.md]: Fixed — moved the YAML frontmatter above the H1 so markdown tooling parses it correctly.
